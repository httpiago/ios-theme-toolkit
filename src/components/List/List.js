import React from 'react'
import PropTypes from 'prop-types'
import { mapClassList } from '../../Utils'

/**
 * List component.
 *
 * @since 0.2.0
 * @see https://httpiago.github.io/ios-theme-toolkit/#/components/List
 */
export default function List({ children, className: aditionalClasses, ...rest }) {

  const prefix = 'list'
  const classes = ''//mapClassList({})

  return (
    <ul
      className={`${prefix} ${classes} ${aditionalClasses}`}
      {...rest}
    >{children}</ul>
  )
}

List.defaultProps = {
  className: '',
}

List.propTypes = {
  /** Aditional classes. */
  className: PropTypes.string,
}


/**
 * List.Item component.
 *
 * @since 0.2.0
 * @see https://httpiago.github.io/ios-theme-toolkit/#/components/List
 */
List.Item = function({ children, href, target, after, media, className: aditionalClasses, ...rest }) {

  const prefix = 'list__item'
  const classes = mapClassList({
    [`${prefix}--link`]: href,
  })

  const content = (
    <React.Fragment>
      {media && (<div className="list__item__media">{media}</div>)}

      <div className="list__item__content">
        {children}

        {after && (<div className="list__item__after">{after}</div>)}
      </div>
    </React.Fragment>
  )

  return (
    <li
      className={`${prefix} ${classes} ${aditionalClasses}`}
      {...rest}
    >
      {href ? (<a href={href} target={target ? target : undefined}>{content}</a>) : content }
    </li>
  )
}

List.Item.defaultProps = {
  className: ''
}

List.Item.propTypes = {
  /** Aditional classes */
  className: PropTypes.string,
  /** Item with link */
  href: PropTypes.string,
  /** link target */
  target: PropTypes.string,
  /** Item media (anything that can be rendered by React, link <img/>) */
  media: PropTypes.node,
  /** Item after (anything that can be rendered by React, link <Badge/> or text) */
  after: PropTypes.node,
}
