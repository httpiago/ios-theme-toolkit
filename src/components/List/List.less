@import '../../customizations.less';

// Default style
.list {
  background-color: #FFFFFF;
  border-color: #c8c7cc;
  border-style: solid;
  border-width: 1px 0;
  position: relative;
  list-style: none;
  font-family: @font-family;
  font-size: @font-size;
  margin: @base-margin 0;
  padding: 0;
}

// Itens dentro da lista
.list__item {
  position: relative;
  box-sizing: border-box;
  min-height: 44px;
  white-space: nowrap;
  text-overflow: ellipsis;
  max-width: 100%;
  display: flex;
  justify-content: space-between;
  align-items: center;
  align-content: center;
}

// Item com link
.list__item--link {
  padding-left: 0;
  padding-right: 0;
}

// Link dentro do item
.list__item--link a {
  min-height: 44px;
  width: 100%;
  padding-left: 0px;
  transition-duration: 0.3s;
  transition-property: background-color;
  text-decoration: none;
  display: flex;
  box-sizing: border-box;
  justify-content: space-between;
  align-items: center;
  position: relative;
  overflow: hidden;
  cursor: pointer;
  color: inherit;
  z-index: 0;

  &:active {
    -webkit-transition-duration: 0s;
    transition-duration: 0s;
    background-color: #d9d9d9;
  }

  & .list__item__content {
    background-image: url("data:image/svg+xml;charset=utf-8,%3Csvg width%3D'8px' height%3D'13px' viewBox%3D'0 0 8 13' xmlns%3D'http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg'%3E%3Cpolygon fill%3D'%23c7c7cc' transform%3D'translate(1.500000%2C 6.500000) rotate(-45.000000) translate(-1.500000%2C -6.500000) ' points%3D'6 11 6 2 4 2 4 9 -3 9 -3 11 5 11'%3E%3C%2Fpolygon%3E%3C%2Fsvg%3E");
    background-position: calc(100% - 15px) center;
    background-repeat: no-repeat;
    background-size: 8px 13px;
    padding-right: 35px;
  }
}

// Item media
.list__item__media {
  display: flex;
  flex-shrink: 0;
  flex-wrap: nowrap;
  align-items: center;
  box-sizing: border-box;
  padding-bottom: 8px;
  padding-top: 7px;
  padding-left: 15px;
}

// Conte√∫do do item
.list__item__content {
  position: relative;
  width: 100%;
  margin-left: 15px;
  padding-top: 8px;
  padding-right: 15px;
  padding-bottom: 8px;
  min-width: 0;
  display: flex;
  justify-content: space-between;
  box-sizing: border-box;
  align-items: center;
  align-self: stretch;
}

.list__item:not(:last-child) .list__item__content::after {
  content: '';
  position: absolute;
  background-color: #c8c7cc;
  display: block;
  z-index: 2;
  top: auto;
  right: 0;
  bottom: 0;
  left: 0px;
  height: 1px;
  width: auto;
}

.list__item__after {
  white-space: nowrap;
  flex-shrink: 0;
  display: flex;
  margin-left: auto;
  color: #8e8e93;
  padding-left: 5px;
}

// Estilo personalizado para listas dentro de um card ou popover
.popover__content .list,
.card__content .list {
  margin: 0px -15px;
}
