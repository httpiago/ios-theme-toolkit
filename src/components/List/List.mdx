---
name: List
route: /components/List
menu: Componentes
---

import { Playground, PropsTable } from 'docz'
import { List, Badge, Toggle, Card } from '../../index.js'
import '../../../dist/styles.css'

# List

To use this component, simply import it into any file and pass the [modifiers](#/#props) in the properties.

```jsx
import { List } from 'ios-theme-toolkit'

export default () => (
  <List>
    <List.Item>1</List.Item>
    <List.Item>2</List.Item>
    <List.Item>3</List.Item>
  </List>
)
```

## Examples

<Playground>
  <span className="desc">sample:</span>
  <List>
    <List.Item>Item 1</List.Item>
    <List.Item>Item 2</List.Item>
    <List.Item>Item 3</List.Item>
  </List>

  <span className="desc">With link:</span>
  <List>
    <List.Item
      href="javascript:console.log('ola1');"
    >
      Item 1
    </List.Item>
    <List.Item
      href="javascript:console.log('ola2');"
    >
      Item 2
    </List.Item>
    <List.Item
      href="javascript:console.log('ola3');"
    >
      Item 3
    </List.Item>
  </List>

  <span className="desc">List with icons inside card:</span>
  <Card>
    <List>
      <List.Item
        media={(<Badge size="big" color="blue">I</Badge>)}
        after="NEW"
      >
        Item 1
      </List.Item>
      <List.Item
        media={(<Badge size="big" color="red">A</Badge>)}
        after={(<Badge color="gray">3</Badge>)}>
        Item 2
      </List.Item>
      <List.Item
        media={(<Badge size="big" color="black">J</Badge>)}
        after={(<Badge size="tiny" color="green"></Badge>)}
      >
        Item 3
      </List.Item>
      <List.Item
        media={(<Badge size="big" color="yellow">M</Badge>)}
        after={(<Toggle color="blue" defaultChecked />)}
      >
        Item 4
      </List.Item>
    </List>
  </Card>

  <span className="desc">With icons and links:</span>
  <List>
    <List.Item
      media={(<Badge size="big" color="blue">I</Badge>)}
      after="NEW"
      href="javascript:console.log('ola1')"
    >
      Item 1
    </List.Item>
    <List.Item
      media={(<Badge size="big" color="red">A</Badge>)}
      after={(<Badge color="gray">3</Badge>)}
      href="javascript:console.log('ola2')"
    >
      Item 2
    </List.Item>
    <List.Item
      media={(<Badge size="big" color="black">J</Badge>)}
      after={(<Badge size="tiny" color="green"></Badge>)}
      href="javascript:console.log('ola3')"
    >
      Item 3
    </List.Item>
  </List>

  <style>{".react-live-preview .desc { display: block; font-weight: bold; margin-top: 20px; }"}</style>
  <style>{'.list__item__media .badge { width: 32px; height: 32px; display: flex; align-items: center; justify-content: center; }'}</style>
</Playground>

## List Props

<PropsTable of={ List } />

All the properties which are not listed above will be transferred to the \<ul class="list"\> tag.

## List.Item Props

<PropsTable of={ List.Item } />

All the properties which are not listed above will be transferred to the \<li class="list__item"\> tag.
