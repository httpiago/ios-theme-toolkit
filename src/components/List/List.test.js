/**
 * TEST FILE
 * Please try writing the tests for all component behaviors,
 * manually call the functions within the class, change the properties...
 *
 * @see https://jestjs.io/docs/en/expect
 * @see https://github.com/airbnb/enzyme/tree/master/docs/api
 */
import React from 'react'
import { shallow } from 'enzyme'
import List from './List'

const COMPONENT = shallow(
  <List className="aditionalClass">
    <List.Item>Item 1</List.Item>
    <List.Item>Item 2</List.Item>
    <List.Item>Item 3</List.Item>
  </List>
)

describe('Test List component', () => {
  it('should render correctly based on last snapshot', () => {
    expect(COMPONENT).toMatchSnapshot()
  })

  it('should has aditional classes', () => {
    expect(COMPONENT.hasClass('aditionalClass')).toBeTruthy()
  })

  it('should pass aditional props to html element', () => {
    COMPONENT.setProps({ 'data-custom-attr': 'yes' })

    expect(COMPONENT.html()).toContain('data-custom-attr="yes"')
  })

  it('should return a list wrapper element', () => {
    expect(['ul', 'ol']).toContain(COMPONENT.type())
  })
  
  it('should render buttons as children correctly', () => {
    expect(COMPONENT.children()).toHaveLength(3)
  })

})

const ITEM = shallow(
  <List.Item
    media={'media content'}
    after={'after badge'}
    className="itemCustomClass"
  >
    <span className="check-content">Item</span>
  </List.Item>
)

describe('Test List.Item component', () => {
  it('should return a list wrapper element', () => {
    expect(ITEM.type()).toEqual('li')
  })

  it('should has aditional classes', () => {
    expect(ITEM.hasClass('itemCustomClass')).toBeTruthy()
  })

  it('should return a link when "href" prop are defined', () => {
    const LINKITEM = shallow(
      <List.Item href="https://www.iagobruno.com/">Item 1</List.Item>
    ).children().first()
    
    expect(LINKITEM.type()).toEqual('a')
    
    expect(LINKITEM.html()).toContain('href="https://www.iagobruno.com/"')
  })

  it('should pass aditional props to html element', () => {
    COMPONENT.setProps({ 'data-custom-attr': 'yes' })

    expect(COMPONENT.html()).toContain('data-custom-attr="yes"')
  })

  it('should render "media" prop if defined', () => {
    expect(ITEM.exists('.list__item__media')).toBeTruthy()
    
    expect(ITEM.find('.list__item__media').text()).toEqual('media content')
    
    // Remove "media" prop and check if the comp update correctly
    ITEM.setProps({ media: undefined })
    
    expect(ITEM.exists('.list__item__media')).toBeFalsy()
  })

  it('should render "after" prop if defined', () => {
    expect(ITEM.exists('.list__item__after')).toBeTruthy()
    
    expect(ITEM.find('.list__item__after').text()).toEqual('after badge')
    
    // Remove "after" prop and check if the comp update correctly
    ITEM.setProps({ after: undefined })
    
    expect(ITEM.exists('.list__item__after')).toBeFalsy()
  })

  it('should render children correctly', () => {
    expect(ITEM.exists('.check-content')).toBeTruthy()
  })

})
