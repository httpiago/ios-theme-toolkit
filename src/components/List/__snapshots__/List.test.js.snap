// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test List component should render correctly based on last snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <List
    className="aditionalClass"
  >
    <Unknown
      className=""
    >
      Item 1
    </Unknown>
    <Unknown
      className=""
    >
      Item 2
    </Unknown>
    <Unknown
      className=""
    >
      Item 3
    </Unknown>
  </List>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Unknown
          className=""
        >
          Item 1
        </Unknown>,
        <Unknown
          className=""
        >
          Item 2
        </Unknown>,
        <Unknown
          className=""
        >
          Item 3
        </Unknown>,
      ],
      "className": "list  aditionalClass",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Item 1",
          "className": "",
        },
        "ref": null,
        "rendered": "Item 1",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Item 2",
          "className": "",
        },
        "ref": null,
        "rendered": "Item 2",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Item 3",
          "className": "",
        },
        "ref": null,
        "rendered": "Item 3",
        "type": [Function],
      },
    ],
    "type": "ul",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Unknown
            className=""
          >
            Item 1
          </Unknown>,
          <Unknown
            className=""
          >
            Item 2
          </Unknown>,
          <Unknown
            className=""
          >
            Item 3
          </Unknown>,
        ],
        "className": "list  aditionalClass",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Item 1",
            "className": "",
          },
          "ref": null,
          "rendered": "Item 1",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Item 2",
            "className": "",
          },
          "ref": null,
          "rendered": "Item 2",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Item 3",
            "className": "",
          },
          "ref": null,
          "rendered": "Item 3",
          "type": [Function],
        },
      ],
      "type": "ul",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
